File: February/feb06/feb06.cs
using System;
using System.Collections.Generic;

class Node {
    public int data;
    public Node left, right;
    public Node(int val) {
        data = val;
        left = right = null;
    }
}

class BinaryTree {
    private static Node ConstructTree(List<int> preorder, List<int> inorder, Dictionary<int, int> inorderMap, ref int preIndex, int inStart, int inEnd) {
        if (inStart > inEnd) return null;

        int rootValue = preorder[preIndex++];
        Node root = new Node(rootValue);
        int rootIndex = inorderMap[rootValue];

        root.left = ConstructTree(preorder, inorder, inorderMap, ref preIndex, inStart, rootIndex - 1);
        root.right = ConstructTree(preorder, inorder, inorderMap, ref preIndex, rootIndex + 1, inEnd);

        return root;
    }

    public static Node BuildTree(List<int> inorder, List<int> preorder) {
        Dictionary<int, int> inorderMap = new Dictionary<int, int>();
        for (int i = 0; i < inorder.Count; i++) {
            inorderMap[inorder[i]] = i;
        }

        int preIndex = 0;
        return ConstructTree(preorder, inorder, inorderMap, ref preIndex, 0, inorder.Count - 1);
    }
}

File: February/feb06/feb06.py
class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None

def construct_tree(preorder, inorder, inorder_map, pre_index, in_start, in_end):
    if in_start > in_end:
        return None

    root_value = preorder[pre_index[0]]
    pre_index[0] += 1
    root = Node(root_value)
    root_index = inorder_map[root_value]

    root.left = construct_tree(preorder, inorder, inorder_map, pre_index, in_start, root_index - 1)
    root.right = construct_tree(preorder, inorder, inorder_map, pre_index, root_index + 1, in_end)

    return root

def build_tree(inorder, preorder):
    inorder_map = {val: idx for idx, val in enumerate(inorder)}
    pre_index = [0]  # Using a list to act as a reference
    return construct_tree(preorder, inorder, inorder_map, pre_index, 0, len(inorder) - 1)
